File: .agents/auditor_1/validate_mermaid.js
const fs = require('fs');
const path = require('path');

const targetDir = 'd:\\Idea_DoAn\\04_Thiet_Ke_Kien_Truc_Diagrams';
const files = [
  '01_Kien_Truc_Tong_Quan.md',
  '02_Sequence_Diagrams.md',
  '03_ERD_Database_Diagram.md',
  '04_Deployment_Diagram.md'
];

let totalBlocks = 0;
let errors = [];

function validateSequenceDiagram(code, file, startLine) {
  const lines = code.split('\n');
  const stack = [];
  lines.forEach((line, idx) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('%%')) return;
    
    // Check block openings
    if (/^(alt|opt|loop|par|critical|rect)\b/i.test(trimmed)) {
      stack.push({ type: trimmed.split(' ')[0], line: startLine + idx });
    } else if (/^else\b/i.test(trimmed)) {
      if (stack.length === 0 || (stack[stack.length - 1].type !== 'alt' && stack[stack.length - 1].type !== 'critical')) {
        errors.push({ file, line: startLine + idx, msg: `Unmatched 'else' without preceding 'alt'` });
      }
    } else if (/^and\b/i.test(trimmed)) {
      if (stack.length === 0 || stack[stack.length - 1].type !== 'par') {
        errors.push({ file, line: startLine + idx, msg: `Unmatched 'and' without preceding 'par'` });
      }
    } else if (/^end\b/i.test(trimmed)) {
      if (stack.length === 0) {
        errors.push({ file, line: startLine + idx, msg: `Unmatched 'end' statement` });
      } else {
        stack.pop();
      }
    }
  });
  if (stack.length > 0) {
    stack.forEach(s => {
      errors.push({ file, line: s.line, msg: `Unclosed block '${s.type}'` });
    });
  }
}

function validateErDiagram(code, file, startLine) {
  const lines = code.split('\n');
  let inEntity = false;
  let currentEntity = '';
  lines.forEach((line, idx) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('%%') || trimmed === 'erDiagram') return;
    
    if (trimmed.endsWith('{')) {
      if (inEntity) {
        errors.push({ file, line: startLine + idx, msg: `Nested entity definition inside ${currentEntity}` });
      }
      inEntity = true;
      currentEntity = trimmed.replace('{', '').trim();
    } else if (trimmed === '}') {
      if (!inEntity) {
        errors.push({ file, line: startLine + idx, msg: `Unmatched closing brace '}' outside entity` });
      }
      inEntity = false;
      currentEntity = '';
    } else if (inEntity) {
      // Must be attribute line: type name [keys] ["comment"]
      const attrMatch = trimmed.match(/^([A-Za-z0-9_()]+)\s+([A-Za-z0-9_]+)(\s+[A-Za-z0-9_,]+)?(\s+"[^"]*")?$/);
      if (!attrMatch) {
        // Some types might have spaces or special characters
        const broadMatch = trimmed.match(/^(\S+)\s+(\S+)/);
        if (!broadMatch) {
          errors.push({ file, line: startLine + idx, msg: `Invalid attribute format in entity ${currentEntity}: ${trimmed}` });
        }
      }
    } else {
      // Must be relationship: ENTITY1 relation ENTITY2 : "comment"
      const relMatch = trimmed.match(/^([A-Za-z0-9_]+)\s+(\|\||\|o|o\||}\||\}|o|\{)(\-\-|\.\.)(\|\||\|o|o\||o\{|\{\||\{\{|\}\})(\s+[A-Za-z0-9_]+)\s*:\s*("[^"]*"|'[^']*'|[A-Za-z0-9_ \-\/()]+)$/);
      const simpleRelMatch = trimmed.match(/^([A-Za-z0-9_]+)\s+[\S]+\s+([A-Za-z0-9_]+)\s*:\s*.*$/);
      if (!relMatch && !simpleRelMatch && !trimmed.startsWith('%%')) {
        errors.push({ file, line: startLine + idx, msg: `Unrecognized erDiagram line: ${trimmed}` });
      }
    }
  });
  if (inEntity) {
    errors.push({ file, line: startLine, msg: `Unclosed entity '${currentEntity}' in erDiagram` });
  }
}

function validateFlowchart(code, file, startLine) {
  const lines = code.split('\n');
  let subgraphStack = [];
  lines.forEach((line, idx) => {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('%%')) return;
    
    if (/^subgraph\b/i.test(trimmed)) {
      subgraphStack.push({ line: startLine + idx, text: trimmed });
    } else if (/^end\b/i.test(trimmed)) {
      if (subgraphStack.length === 0) {
        errors.push({ file, line: startLine + idx, msg: `Unmatched 'end' in flowchart` });
      } else {
        subgraphStack.pop();
      }
    }
  });
  if (subgraphStack.length > 0) {
    subgraphStack.forEach(s => {
      errors.push({ file, line: s.line, msg: `Unclosed subgraph '${s.text}'` });
    });
  }
}

files.forEach(filename => {
  const filePath = path.join(targetDir, filename);
  const content = fs.readFileSync(filePath, 'utf8');
  
  const mermaidRegex = /```mermaid\r?\n([\s\S]*?)```/g;
  let mMatch;
  let bIdx = 0;
  while ((mMatch = mermaidRegex.exec(content)) !== null) {
    bIdx++;
    totalBlocks++;
    const startLine = content.substring(0, mMatch.index).split('\n').length;
    const diagramCode = mMatch[1].trim();
    const firstLine = diagramCode.split('\n')[0].trim();
    
    console.log(`[${filename}] Block #${bIdx} at Line ${startLine}: Type='${firstLine}' (${diagramCode.split('\n').length} lines)`);
    
    if (firstLine.startsWith('sequenceDiagram')) {
      validateSequenceDiagram(diagramCode, filename, startLine);
    } else if (firstLine.startsWith('erDiagram')) {
      validateErDiagram(diagramCode, filename, startLine);
    } else if (firstLine.startsWith('flowchart') || firstLine.startsWith('graph')) {
      validateFlowchart(diagramCode, filename, startLine);
    }
  }
});

console.log(`\nTotal Mermaid Blocks Checked: ${totalBlocks}`);
console.log(`Total Syntax / Structural Errors: ${errors.length}`);
if (errors.length > 0) {
  console.log('Errors:', JSON.stringify(errors, null, 2));
}

File: .agents/reviewer_1/validate_mermaid.py
import os
import re

files = [
    r'd:\Idea_DoAn\04_Thiet_Ke_Kien_Truc_Diagrams\01_Kien_Truc_Tong_Quan.md',
    r'd:\Idea_DoAn\04_Thiet_Ke_Kien_Truc_Diagrams\02_Sequence_Diagrams.md',
    r'd:\Idea_DoAn\04_Thiet_Ke_Kien_Truc_Diagrams\03_ERD_Database_Diagram.md',
    r'd:\Idea_DoAn\04_Thiet_Ke_Kien_Truc_Diagrams\04_Deployment_Diagram.md'
]

def check_sequence(code, filename, block_idx):
    lines = code.strip().split('\n')
    stack = []
    activations = {}
    errors = []
    
    for line_no, raw_line in enumerate(lines, 1):
        line = raw_line.strip()
        if not line or line.startswith('%%'):
            continue
            
        first_word = line.split()[0] if line.split() else ''
        if first_word in ['alt', 'opt', 'par', 'loop', 'rect', 'critical', 'break']:
            stack.append((first_word, line_no))
        elif first_word == 'else':
            if not stack or stack[-1][0] not in ['alt', 'critical']:
                errors.append(f'Line {line_no}: "else" without matching alt/critical')
        elif first_word == 'and':
            if not stack or stack[-1][0] != 'par':
                errors.append(f'Line {line_no}: "and" without matching par')
        elif first_word == 'end':
            if not stack:
                errors.append(f'Line {line_no}: "end" without matching block start')
            else:
                stack.pop()
                
        if line.startswith('activate '):
            part = line.split()[1]
            activations[part] = activations.get(part, 0) + 1
        elif line.startswith('deactivate '):
            part = line.split()[1]
            activations[part] = activations.get(part, 0) - 1
            if activations[part] < 0:
                errors.append(f'Line {line_no}: deactivate {part} without corresponding activate')

    if stack:
        for s, lno in stack:
            errors.append(f'Unclosed block "{s}" started at line {lno}')
    for part, count in activations.items():
        if count > 0:
            errors.append(f'Participant "{part}" activated {count} more times than deactivated')
            
    return errors

def check_flowchart(code, filename, block_idx):
    lines = code.strip().split('\n')
    stack = []
    errors = []
    for line_no, raw_line in enumerate(lines, 1):
        line = raw_line.strip()
        if not line or line.startswith('%%'):
            continue
        if line.startswith('subgraph '):
            stack.append(('subgraph', line_no))
        elif line == 'end':
            if not stack:
                errors.append(f'Line {line_no}: "end" without matching subgraph')
            else:
                stack.pop()
    if stack:
        for s, lno in stack:
            errors.append(f'Unclosed subgraph started at line {lno}')
    return errors

def check_er(code, filename, block_idx):
    lines = code.strip().split('\n')
    errors = []
    in_entity = False
    entity_start = 0
    entity_name = ''
    for line_no, raw_line in enumerate(lines, 1):
        line = raw_line.strip()
        if not line or line.startswith('%%'):
            continue
        if '{' in line and not in_entity:
            in_entity = True
            entity_start = line_no
            entity_name = line.split('{')[0].strip()
        elif line == '}' and in_entity:
            in_entity = False
        elif '}' in line and not in_entity:
            errors.append(f'Line {line_no}: closing brace without opening entity')
    if in_entity:
        errors.append(f'Unclosed entity {entity_name} started at line {entity_start}')
    return errors

all_errors = []
for filepath in files:
    fname = os.path.basename(filepath)
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    blocks = re.findall(r'```mermaid(.*?)```', content, re.DOTALL)
    for i, b in enumerate(blocks):
        first_line = b.strip().split('\n')[0].strip()
        diagram_type = first_line.split()[0]
        if diagram_type in ['flowchart', 'graph']:
            errs = check_flowchart(b, fname, i+1)
        elif diagram_type == 'sequenceDiagram':
            errs = check_sequence(b, fname, i+1)
        elif diagram_type == 'erDiagram':
            errs = check_er(b, fname, i+1)
        else:
            errs = [f'Unknown diagram type: {diagram_type}']
        
        if errs:
            for e in errs:
                all_errors.append(f'[{fname} Block {i+1} ({diagram_type})]: {e}')
        else:
            print(f'OK: [{fname} Block {i+1} ({diagram_type})]')

if all_errors:
    print('\nFound structural issues:')
    for e in all_errors:
        print('  - ' + e)
else:
    print('\nAll 22 mermaid blocks passed structural syntax analysis!')
